### DISCUSSION TITLE
Help migrate NuGet from XF to MAUI

### DISCUSSION THREAD

**Discussion by IeuanWalker** [CommunityToolkit/Maui#275] [General]
Hi, hoping someone can point me in the right direction.

Looking at migrating some of my NuGets to Maui, but struggling atm.

I've stripped down the project to the bare minimum _(removed OS specific renders - will be adding back once the basic functionality works)_. I've copied the toolkit .csproj, but getting this error not sure where the issue is - 

```
1>D:\WS\GitHub\Xamarin.Forms.CustomSwitch\Src\Switch\Microsoft.Maui.Controls.SourceGen\Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator\CustomSwitch.xaml.sg.cs(5,46,5,46): error CS1001: Identifier expected
1>Done building project "Switch.csproj" -- FAILED.
```

Here is the branch I'm currently working on - https://github.com/IeuanWalker/Xamarin.Forms.CustomSwitch/tree/MauiMigration

**Comment by pictos** [comment]
@IeuanWalker if you remove all `x:Name` from xaml and C# code-behind, the error disappears?

**Reply by IeuanWalker** [reply]
@pictos got a basic version working now. I forgot to update the xamarin reference in the control -
```xmlns="http://xamarin.com/schemas/2014/forms"```

The one bit I'm concerned about now is how to handle the platform-specific renderers. I deleted them to get something basic working as a lot of the namespaces weren't being found. 

The purpose of the renderer is to make the specific platforms treat the control as a native switch _(for accessibility)_. Is this something that can be updated to handlers?

- Android - https://github.com/IeuanWalker/Xamarin.Forms.CustomSwitch/blob/master/Src/Switch.Android/CustomSwitchRenderer.cs
- iOS - https://github.com/IeuanWalker/Xamarin.Forms.CustomSwitch/blob/master/Src/Switch.iOS/CustomSwitchRenderer.cs

**Reply by pictos** [reply]
You can take a look on my work porting the XCT PopupRenderer to MCT PopupHandler [right here](https://github.com/CommunityToolkit/Maui/tree/pj/popup-to-handler). Also Shane and I have an amazing session do the porting for Android, you find it [here](https://www.twitch.tv/videos/1215074147)

**Reply by IeuanWalker** [reply]
Thanks @pictos, iv managed to get a renderer working on iOS. But the android renderer isn't working.
The code in the renderer is exactly the same as XF version, just updated namespaces.

All the renderer does is support multiple corner radiuses.

On Android, it's not loading anything on the page, but the debugger does enter the renderer

https://github.com/IeuanWalker/Xamarin.Forms.CustomSwitch/commit/06ecbf0b921dfb185a921c78c3ce7780ea4ae904

**Reply by pictos** [reply]
I see that you're using the compatibilityRenderer, I know there are dragons in that area... If you change everything to use handler that should work

**Reply by IeuanWalker** [reply]
@pictos Ye looked at doing that originally, but couldn't find a handler/ interface for a `Frame`. So thought i try and get the renderers working then see about updating it to the new approach.

I've looked in here but couldn't find it - https://github.com/dotnet/maui/tree/main/src/Core/src/Handlers

Any ideas, on how to use a handler for a frame?


Found this, suggesting there is one somewhere - https://github.com/dotnet/maui/wiki/status#-frame

**Reply by pictos** [reply]
How about using a ContentView instead of Frame, now you can shadow in any view, maybe CornerRadius as well

**Reply by IeuanWalker** [reply]
@pictos using a ContentView instead, the control doesn't display. _(without any other changes, just a straight swap with the custom frame)_

**Reply by pictos** [reply]
@IeuanWalker if you set a HeightRequest to the View does it shows? Because if your code is shared and works on iOS that should work on Android as well.

**Reply by IeuanWalker** [reply]
@pictos nope, the height request is already being set

**Reply by pictos** [reply]
@IeuanWalker I tried to run on my end, but I got this error on Android and can't work around it, is it familiar to you?


<details><summary>Show/Hide Visual Studio info</summary>

```
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/colorPrimary (aka Switch.Switch:attr/colorPrimary)' not found.
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/colorPrimaryDark (aka Switch.Switch:attr/colorPrimaryDark)' not found.
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/colorAccent (aka Switch.Switch:attr/colorAccent)' not found.
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/materialButtonStyle (aka Switch.Switch:attr/materialButtonStyle)' not found.
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/windowActionBar (aka Switch.Switch:attr/windowActionBar)' not found.
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/windowNoTitle (aka Switch.Switch:attr/windowNoTitle)' not found.
1>C:\Program Files\dotnet\packs\Microsoft.iOS.Windows.Sdk\15.2.200-preview.12.4\tools\msbuild\iOS\Xamarin.Messaging.Apple.targets(31,3): warning : There is no available connection to the Mac. Task 'VerifyXcodeVersion' will not be executed
1>C:\Users\Jesus\source\repos\Xamarin.Forms.CustomSwitch\Src\Switch\obj\Debug\net6.0-android\lp\68\jl\res\values\styles.xml(2): error APT2260: style attribute 'attr/windowActionModeOverlay (aka Switch.Switch:attr/windowActionModeOverlay)' not found.
1>C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\31.0.200-preview.12.2\tools\Xamarin.Android.Aapt2.targets(157,3): error APT2260: resource style/Widget.MaterialComponents.Button.Icon (aka Switch.Switch:style/Widget.MaterialComponents.Button.Icon) not found.
1>C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\31.0.200-preview.12.2\tools\Xamarin.Android.Aapt2.targets(157,3): error APT2062: failed linking references.
```
</details>

**Reply by IeuanWalker** [reply]
@pictos strange, no I just get warnings about some obsolete methods 
![image](https://user-images.githubusercontent.com/6544051/152765185-03288582-2021-43a9-ae37-23d1be89eefb.png)

The MAUI sample app is newly created and doesn't have a styles.xml file.

**Reply by VladislavAntonyuk** [reply]
@pictos you probably need to rebuild it. looks like your Android Resource.Designer.cs is missed

**Reply by IeuanWalker** [reply]
@pictos @VladislavAntonyuk 
Could one of you help with migrating this project - https://github.com/IeuanWalker/Xamarin.Forms.StateButton

I've tried migrating it to Maui using the handler approach, but it's just not working - https://github.com/IeuanWalker/Maui.StateButton

The control is based on the Border control, and it needs to go down to each platform and implement clicked, pressed and released events. And then call the methods in the shared control to handle each event.

**Reply by VladislavAntonyuk** [reply]
I haven’t looked in details but I think you incorrectly registered handler. There is method AddHandler instead of AddTransient.

**Reply by IeuanWalker** [reply]
@VladislavAntonyuk  Ahh ok thanks, I copied that from here lol - https://github.com/jsuarezruiz/ways-create-netmaui-controls/blob/44dfb1c0a52c67c7238c16ad1ca787b52b6d3bfa/src/CreateControls/Controls/Handler/AppHostBuilderExtensions.cs#L31

I've updated my code, but it hasn't changed anything :/

**Reply by pictos** [reply]
@IeuanWalker what's the Xamarin version of the StateButton? Basically you don't need to use a Handler to access the platform-specific code, you can use the mapper or PlatformBehavior or even the propertyChanged delegate. You can find more info [here](https://devblogs.microsoft.com/dotnet/customizing-dotnet-maui-controls/)

**Reply by IeuanWalker** [reply]
@pictos this is the xamarin control - https://github.com/IeuanWalker/Xamarin.Forms.StateButton

The issue is I don't want the new events to be in a mapper, as they are used within the control itself, and don't want to make them available externally.

And using a behaviour or a mapper I couldn't see how you override methods on each of the platforms -

- Android OnKeyUp - 
    - https://github.com/IeuanWalker/Xamarin.Forms.StateButton/blob/e56a3c037ed8dbd7e5235a0987df200aea13810c/Scr/StateButton.Android/StateButtonRenderer.cs#L95
- iOS (TouchesMoved, TouchesBegan, TouchesCancelled)
    -  https://github.com/IeuanWalker/Xamarin.Forms.StateButton/blob/e56a3c037ed8dbd7e5235a0987df200aea13810c/Scr/StateButton.iOS/StateButtonRenderer.cs#L44

**Reply by IeuanWalker** [reply]
@pictos I got the control working on Android using a behaviour, only thing left on android is overriding the `OnKeyUp` method.

The behavior code is on this branch - https://github.com/IeuanWalker/Maui.StateButton/tree/TestBehavior

For iOS not sure how id override the touch methods `(TouchesMoved, TouchesBegan, TouchesCancelled)`

**Reply by pictos** [reply]
@IeuanWalker you just need to create your own type that inherits from `ContentViewGroup`(android) and `ContentView` iOS, then you will be able to override the methods that you want. For example:
![image](https://user-images.githubusercontent.com/20712372/183786695-8e8ada14-d3c8-44f7-9afd-d472ec2efb2e.png)



That should do the job (:

**Reply by IeuanWalker** [reply]
@pictos thanks, is there a way to do that with a platform behavior?

**Reply by pictos** [reply]
I don't think so, because you will need to provide a new `PlatformElement` to the control... The behavior doesn't suppor that...

**Reply by IeuanWalker** [reply]
@pictos I've set up everything how I think the handler should be, but the control just isn't rendering.

The android `ConnectHandler` is being reached when I debug it.

Commenting out `handlers.AddHandler(typeof(StateButton), typeof(StateButtonHandler));` then the border renders as expected, but when using the handler nothing is rendered.

If you have time could you take a look, all my handler code is on the master branch _(only looking at android atm)_ - https://github.com/IeuanWalker/Maui.StateButton

**Reply by pictos** [reply]
Maybe you need to call the AddView method manually. I remember that I had to do it while ago... But didn't investigate futher to see if was a bug or something we need to do

**Reply by IeuanWalker** [reply]
@pictos copied this from one of the Maui controls -
```csharp
		if (VirtualView.PresentedContent is IView view)
		{
			var platformContent = view.ToPlatform(MauiContext);
			// Make sure we don't belong to a previous parent
			platformContent.RemoveFromParent();
			PlatformView.AddView(platformContent);
		}
```

`PlatformView.AddView(platformContent);` is being hit, but the control still isn't being rendered.

_same thing happening on iOS_

**Comment by CliffAgius** [comment]
Closed as answered
